### ISSUE TITLE
vhost core dump during bdev create and VM start up

### ISSUE THREAD

**Issue by mhae** [spdk/spdk#680]
I'm running a stress test scenario where a script is constantly adding and removing devices. At the same time, I start a VM (the test script is started beforehand).
Quite frequently, I get a vhost core dump during add_vhost_scsi_lun.

```
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `app/vhost/vhost -S /var/tmp -s 1024 -m 0x03'.
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x000055cb55eee8b1 in spdk_vhost_scsi_session_add_tgt (vdev=0x20002cd1dfc0, vsession=0x20002cd13780, ctx=0x5) at vhost_scsi.c:912
912             svsession->scsi_dev_state[scsi_tgt_num] = svsession->svdev->scsi_dev_state[scsi_tgt_num];
[Current thread is 1 (Thread 0x7f6e7e4e2980 (LWP 19530))]
```
Note, the line number is one off compared to what's on master since I added a printf to dump the scsi_tgt_num.

I understand that this may be a corner case issue. However, I don't think we can preclude this from happening in real world production environments.

## Expected Behavior
No core dump. Adding a target while a VM is coming up should not cause an abort.

## Current Behavior
core dump

## Steps to Reproduce

1) Start the generic vhost
```
app/vhost/vhost -S /var/tmp -s 1024 -m 0x03
```

2) Add the scsi controller
```
./scripts/rpc.py construct_vhost_scsi_controller --cpumask 0x1 vhost.0
```
3) Run the following stress test script
```
for i in {0..7} ; do
        echo $i
  ./scripts/rpc.py construct_malloc_bdev 1 512 -b Malloc$i
  ./scripts/rpc.py add_vhost_scsi_lun vhost.0 $i Malloc$i
  sleep 0.1
done
for i in {0..7} ; do
  ./scripts/rpc.py remove_vhost_scsi_target vhost.0 $i
  ./scripts/rpc.py delete_malloc_bdev Malloc$i
done
done
```

4) Run the following KVM command line:
```
kvm -enable-kvm -machine pc-i440fx-2.1,accel=kvm,usb=off -smp 1 -cpu SandyBridge -m 512M -object memory-backend-file,id=mem,size=512M,mem-path=/dev/hugepages/vm,share=on -numa node,memdev=mem -drive file=/home/mhae/min.qcow2,if=none,id=disk -device ide-hd,drive=disk,bootindex=0 -chardev socket,id=char0,path=/var/tmp/vhost.0 -device vhost-user-scsi-pci,id=scsi0,chardev=char0 --curses
```

## Context (Environment including OS version, SPDK version, etc.)
mhae@mhae-z800:~/src/spdk$ uname -a
Linux mhae-z800 4.13.0-46-generic #51-Ubuntu SMP Tue Jun 12 12:36:29 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

SPDK version is top of tree (master).

**Comment by jimharris** [comment]
Assigning to @darsto for now.  Also tagging for next bug scrub - maybe @darsto will decide to assign it to someone else.  :-)

The test scenario is certainly valid.  I'm marking it as a bug mainly because I don't think we have a test for this currently and expect we'll have no problem reproducing the problem.

**Comment by pawelwod** [comment]
I think #687 is duplicate. Lets track it there.

**Comment by peluse** [comment]
@mhae can you please try testing with https://review.gerrithub.io/c/spdk/spdk/+/447060/ ?

**Comment by mhae** [comment]
@Paul, the link you've referenced is the test script which is pretty much what I've used. Was your intention for me to verify that the test script works or the actual fix? I did test the fix as described in https://github.com/spdk/spdk/issues/687 and it worked fine.
If your intention is to really verify the test, then I can surely do that. However, I was wondering if they run out of the box or if I have to do some additional setup.
